✨ feat : 채팅방 목록 조회 기능 구현

BE12-OZ opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #119 
- 작업 요약: 채팅방 목록 조회 기능 구현 
(이전 완료된 작업이었으나 우선순위 판단 미스로 인해 웹소켓 메시지 수정/삭제 기능 구현을 진행 -> 피드백 받은 후 수정/보완하여 PR요청)

## 📄 상세 내용
- [x] 채팅방 목록 조회 API 구현:
- 사용자가 속한 모든 채팅방의 목록을 조회하는 API(GET /api/v1/chat/rooms)를 구현
- 각 채팅방 정보에는 스터디 그룹 ID, 그룹명, 마지막 메시지, 안 읽은 메시지 수가 포함

- ChatRoomService 추가 (apps/chat/services/chat_room_service.py)
- 채팅방 목록 조회를 위한 비즈니스 로직을 서비스 계층으로 분리
- ORM의 annotate, Subquery 기능을 활용하여, 각 채팅방의 마지막 메시지와 안 읽은 메시지 수를 단일 쿼리셋으로 효율적 계산
   이를 통해 N+1 문제를 방지
- Coalesce 함수를 사용하여, 사용자의 읽은 기록이 없는 케이스에서도 unread_count가 null이 아닌 0으로 반환되도록 안정성 높힘

- ChatRoomSerializer 및 ChatRoomListView 추가
- 서비스 계층에서 처리된 데이터를 API 응답 형식에 맞게 직렬화하는 ChatRoomSerializer를 추가
- ChatRoomListView를 구현하여 API 요청을 처리하고, Spec 문서를 extend_schema로 상세히 작성

- [x] 테스트 코드 강화 (apps/chat/tests/test_chat_room_api.py)
- 기본적인 API 성공 및 인증 실패 테스트 외에, 다음과 같은 다양한 엣지 케이스를 검증하는 테스트 코드를 추가
- 채팅방에 메시지가 하나도 없는 경우
- 사용자가 아직 읽은 메시지가 하나도 없는 경우
- 사용자가 일부 메시지만 읽은 경우
- 사용자가 모든 메시지를 읽은 경우

- [x] mypy 타입 에러 해결
- 신규로 작성된 모든 코드에 대해 mypy 타입 검사를 통과하도록 수정하여 코드의 안정성과 유지보수성을 향상

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

BE12-OZ committed:
✨ feat: 채팅방 목록 조회 API 구현

- 채팅방 목록 조회 API 엔드포인트(/api/v1/chat/rooms)를 구현했습니다.
- API는 사용자가 속한 스터디 그룹(채팅방)의 ID, 이름, 마지막 메시지, 안 읽은 메시지 수를 반환합니다.
- 비즈니스 로직을 ChatRoomService로 분리하여 관리하고, ORM의 annotate와 Subquery를 사용하여 N+1 쿼리 문제를 방지하고 성능을 최적화했습니다.
- Coalesce 함수를 적용하여 사용자가 메시지를 처음 읽는 경우에도 안 읽은 메시지 수가 정확히 계산되도록 엣지 케이스를 처리했습니다.
- 기능 요구사항을 검증하기 위한 단위 테스트 코드를 작성했으며, 메시지가 없는 경우, 모든 메시지를 읽은 경우 등의 엣지 케이스를 포함했습니다.
- mypy 타입 검사를 통과하여 코드의 정적 안정성을 확보했습니다.

Meoyoug commented on apps/chat/services/chat_room_service.py:
```diff
@@ -0,0 +1,46 @@
+
+from apps.chat.models import ChatMessage, LastReadMessage
+from apps.studies.models.groups import StudyGroup
+from apps.users.models import User
+
+
+class ChatRoomService:
+    @staticmethod
+    def get_chat_rooms_for_user(user: User) -> Iterable[Dict[str, Any]]:
+        # 사용자가 속한 스터디 그룹 목록 조회
+        study_groups = StudyGroup.objects.filter(members__user=user)
+
+        # 각 그룹의 마지막 메시지 서브쿼리
+        last_message_subquery = (
+            ChatMessage.objects.filter(study_group_id=OuterRef("pk")).order_by("-created_at").values("content")[:1]
+        )
+
+        # 사용자의 마지막 읽은 메시지 ID 서브쿼리 (존재하지 않을 경우를 대비)
+        last_read_message_id_subquery = LastReadMessage.objects.filter(study_group_id=OuterRef("pk"), user=user).values(
+            "message_id"
+        )
+
+        # 안 읽은 메시지 수 계산 서브쿼리
+        unread_count_subquery = (
+            ChatMessage.objects.filter(
+                study_group_id=OuterRef("pk"),
+                id__gt=Coalesce(Subquery(last_read_message_id_subquery), 0),
+            )
+            .values("study_group_id")
+            .annotate(count=Count("id"))
+            .values("count")
+        )
+
+        # 스터디 그룹 정보와 함께 마지막 메시지, 안 읽은 메시지 수 조합
+        chat_rooms = study_groups.annotate(
+            last_message=Subquery(last_message_subquery),
+            unread_count=Coalesce(Subquery(unread_count_subquery, output_field=models.IntegerField()), 0),
+        ).values("id", "name", "last_message", "unread_count")
+
+        return chat_rooms
```
<img width="386" height="456" alt="Image" src="https://github.com/user-attachments/assets/5c7bea17-95cc-4cb1-b3a8-021ef842db28" />

마지막 메시지의 내용만 추가해주셧는데
마지막 메시지의 작성자정보(닉네임), 내용, 전송일시도 추가되어야합니다!

BE12-OZ replied:
코치님 리뷰 감사드립니다. 체크해서 수정/보완 하겠습니다 ! bb

BE12-OZ committed:
♻️ refactor: 채팅방 목록 API 코드 리뷰 피드백 반영 및 mypy 오류 수정

- ChatRoomService: 마지막 메시지에 작성자 닉네임, 전송 일시 포함하도록 수정.
- ChatRoomSerializer: 마지막 메시지 구조 변경에 맞춰 중첩 Serializer 적용.
- test_chat_room_api.py: 변경된 last_message 구조에 맞춰 테스트 코드 업데이트.
- mypy: 발생한 모든 타입 오류 해결.
